pkg/interop/types.go

```diff
@@ -35,3 +37,40 @@ type PublicKey []byte
 // Interface represents interop interface type which is needed for
 // transparent handling of VM-internal types (e.g. storage.Context).
 type Interface interface{}
+
+// Equals compares Hash160 with the provided stackitem using EQUAL opcode.
+func (a Hash160) Equals(b interface{}) bool {
+	ha := interface{}(a)
+	return bytesEquals(ha, b)
+}
+
+// Equals compares Hash256 with the provided stackitem using EQUAL opcode.
+func (a Hash256) Equals(b interface{}) bool {
+	ha := interface{}(a)
+	return bytesEquals(ha, b)
+}
+
+// Equals compares PublicKey with the provided stackitem using EQUAL opcode.
+func (a PublicKey) Equals(b interface{}) bool {
+	ha := interface{}(a)
+	return bytesEquals(ha, b)
+}
+
+// Equals compares Signature with the provided stackitem using EQUAL opcode.
+func (a Signature) Equals(b interface{}) bool {
+	ha := interface{}(a)
+	return bytesEquals(ha, b)
+}
+
+// bytesEquals is an internal helper function allowed to compare types with underlying
+// byte array.
+func bytesEquals(a interface{}, b interface{}) bool {
+	ret := a == nil && b == nil
+	if !ret {
+		ret = a != nil && b != nil
+		if ret {
+			ret = neogointernal.Opcode2("EQUAL", a.(string), b.(string)).(bool)
+		}
+	}
+	return ret
```

Comment:
Have you checked the resulting bytecode? Maybe it'll be somewhat more efficient with
`return (a == nil && b == nil) || (a != nil && b != nil && neogointernal.Opcode2("EQUAL", a.(string), b.(string)).(bool))`
? At least this eliminates a local variable.

Reply:
Yes, this code is more efficient both in terms of prog length and prog cost. The initial program for `both nil` test is:
```
INDEX    OPCODE      PARAMETER
0        INITSLOT    4 local, 0 arg    <<
3        PUSHNULL
4        STLOC0
5        PUSHNULL
6        STLOC1
7        LDLOC0
8        STLOC2
9        LDLOC2
10       ISNULL
11       JMPIFNOT    17 (6/06)
13       LDLOC1
14       ISNULL
15       JMP         20 (5/05)
17       PUSH0
18       CONVERT     Boolean (20)
20       STLOC3
21       LDLOC3
22       NOT
23       JMPIFNOT    59 (36/24)
25       LDLOC2
26       ISNULL
27       JMPIF       34 (7/07)
29       LDLOC1
30       ISNULL
31       NOT
32       JMP         37 (5/05)
34       PUSH0
35       CONVERT     Boolean (20)
37       STLOC3
38       LDLOC3
39       JMPIFNOT    59 (20/14)
41       LDLOC2
42       DUP
43       ISTYPE      ByteString (28)
45       JMPIF       49 (4/04)
47       CONVERT     ByteString (28)
49       LDLOC1
50       DUP
51       ISTYPE      ByteString (28)
53       JMPIF       57 (4/04)
55       CONVERT     ByteString (28)
57       EQUAL
58       STLOC3
59       LDLOC3
60       RET
```
The updated program for the same test is:
```
INDEX    OPCODE      PARAMETER
0        INITSLOT    3 local, 0 arg    <<
3        PUSHNULL
4        STLOC0
5        PUSHNULL
6        STLOC1
7        LDLOC0
8        STLOC2
9        LDLOC2
10       ISNULL
11       JMPIFNOT    17 (6/06)
13       LDLOC1
14       ISNULL
15       JMP         20 (5/05)
17       PUSH0
18       CONVERT     Boolean (20)
20       JMPIF       54 (34/22)
22       LDLOC2
23       ISNULL
24       JMPIF       49 (25/19)
26       LDLOC1
27       ISNULL
28       JMPIF       49 (21/15)
30       LDLOC2
31       DUP
32       ISTYPE      ByteString (28)
34       JMPIF       38 (4/04)
36       CONVERT     ByteString (28)
38       LDLOC1
39       DUP
40       ISTYPE      ByteString (28)
42       JMPIF       46 (4/04)
44       CONVERT     ByteString (28)
46       EQUAL
47       JMP         52 (5/05)
49       PUSH0
50       CONVERT     Boolean (20)
52       JMP         57 (5/05)
54       PUSH1
55       CONVERT     Boolean (20)
57       RET
```

So we're saving one slot local variable.